### ISSUE TITLE
4.0.x doesn't support Redux DevTools Chrome Extension

### ISSUE THREAD

**Issue by johanzhou8** [angular-redux/ng-redux#196]
Redux DevTools Chrome Extension doesn't work after I upgrade the ng-redux to 4.0.x.
The console shows `Error: Actions must be plain objects. Use custom middleware for async actions.` even though I've already used the `thunk` middleware like this 
```
$ngReduxProvider.createStoreWith(rootReducer, [thunk], [window.__REDUX_DEVTOOLS_EXTENSION__()]);
```  
The ng-redux library works well if I remove the storeEnhancer `[window.__REDUX_DEVTOOLS_EXTENSION__()]`.

I can reproduce the issue in the example project(https://github.com/johanzhou8/ng-redux/tree/master/examples/async-immutable) too.
Steps to reproduce the issue:
1. download the example project
1. install the dependencies. `npm install`
1. upgrade the ng-redux. `npm i ng-redux@4.0.3`
1. use the redux devtool storeEnhancer in the index.js file. `$ngReduxProvider.createStoreWith(rootReducer, [thunk], [window.__REDUX_DEVTOOLS_EXTENSION__()]);`
1. start the server. `npm start`
1. look at the error message in the console

The workaround for me is downgrading to the older version. `npm i ng-redux@3.5.2`.
Thanks.

**Comment by GuyPie** [comment]
This happens to me as well, except in my case I wasn't seeing any error, just that redux-observable actions weren't working when the devtools were included. Also, I noticed redux-immutable-state-invariant complained about how my reducer was mutating my state even though I was not, this is also not happening in version 3.5.2

**Comment by bjohnso5** [comment]
I get this same error with 4.0.x and `redux-thunk`. 3.5.2 works as expected.

**Comment by AntJanus** [comment]
Has anyone been able to figure out where the issue happens? 

The 4.0.X release had that whole `provideStore` feature. I wonder if that's causing it. I'm one of the maintainers but I barely have time. If any of you can contribute with a PR or further debugging info, that'd be immensely helpful.

**Comment by AntJanus** [comment]
Shit, I just realized what it is. Alright, I'd love some suggestions here.

@bjohnso5 the reason for the issue is the new `provideStore` mechanism. It wraps the provided store and subscribes to it to push changes from the AngularJS store to the provided store. This means that passing functions will cause an error.

Hmmm...

@derrickpelletier you worked a little bit on the provided store feature, do you have any ideas about how to resolve this?

**Comment by derrickpelletier** [comment]
Ah, i'll try to dig into this a bit later today or this weekend (away for holidays).

My gut is that the issue comes from the fact that both stores need awareness of the same middleware, and this isn't reflected in the docs right now...

For example, here's how i'm using this in production and I have both dev tools, logger, and thunk working (and a working branch of sagas).

```
const middleware = [reduxThunk]
if (process.env.NODE_ENV === 'development') {
  middleware.push(createLogger({
    level: 'info',
    collapsed: true
  }))
}

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
const enhancers = composeEnhancers(applyMiddleware(...middleware))
const store = createStore(reducers, enhancers)

angular
  // other stuff...
  .config(($ngReduxProvider) => {
    $ngReduxProvider.provideStore(store, [reduxThunk])
  })
```

So in this case, the ng-redux store use thunk compatible for the purpose of passing actions up to the true store, but the logger and dev tools only exist on the true store.

**Comment by derrickpelletier** [comment]
@AntJanus My previous comment is still a relevant problem, but not the issue being discussed here.

I was digging through some commits tonight to see where this was introduced. From what I can tell, this PR introduced the bug way back in september: https://github.com/angular-redux/ng-redux/pull/166

The reason the bug never appeared until now is because `ng-redux` was never published to npm after the `3.5.2` release so this change didn't propagate until it ended up in `4.x`.

**Comment by NadiykaS** [comment]
I have the same problem. Shall i use 3.5.2 version in order to get it fixed?

**Comment by derrickpelletier** [comment]
@NadiykaS, yeah unfortunately. 4.x only introduces the provideStore functionality, so i don't think you'll miss out on too much.

**Comment by ProdigySim** [comment]
Why does the provideStore functionality have to copy the store? If it's just to add in middleware, we could try a [dynamic-middleware](https://github.com/pofigizm/redux-dynamic-middlewares) solution which would allow ng-redux middleware to be added to the store after its creation.

**Comment by villesau** [comment]
Try this approach which allows you to create the whole store in a callback: https://github.com/angular-redux/ng-redux/pull/218
